modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotCheckProcess.java

```diff
@@ -0,0 +1,503 @@
+
+        phase1CheckMetas = new DistributedProcess<>(kctx, SNAPSHOT_CHECK_METAS, this::prepareAndCheckMetas,
+            this::reducePreparationAndMetasCheck);
+
+        phase2PartsHashes = new DistributedProcess<>(kctx, SNAPSHOT_VALIDATE_PARTS, this::validateParts,
+            this::reduceValidatePartsAndFinish);
+
+        kctx.event().addLocalEventListener((evt) -> nodeLeft(((DiscoveryEvent)evt).eventNode()), EVT_NODE_FAILED, EVT_NODE_LEFT);
+    }
+
+    /** */
+    Map<String, SnapshotCheckProcessRequest> requests() {
+        return Collections.unmodifiableMap(requests);
+    }
+
+    /**
+     * Stops the process with the passed exception.
+     *
+     * @param th The interrupt reason.
+     * @param rqFilter If not {@code null}, used to filter which requests/process to stop. If {@code null}, stops all the validations.
+     */
+    void interrupt(Throwable th, @Nullable Function<SnapshotCheckProcessRequest, Boolean> rqFilter) {
+        requests.values().forEach(rq -> {
+            if (rqFilter == null || rqFilter.apply(rq)) {
+                rq.error(th);
+
+                clean(rq.requestId(), th, null, null);
+            }
+        });
+    }
+
+    /** Stops the related validation if the node is a mandatory one. */
+    private void nodeLeft(ClusterNode node) {
+        if (node.isClient() || requests.isEmpty())
+            return;
+
+        interrupt(
+            new ClusterTopologyCheckedException("Snapshot checking stopped. A node left the cluster: " + node + '.'),
+            rq -> rq.nodes().contains(node.id())
+        );
+    }
+
+    /** Phase 2 and process finish. */
+    private IgniteInternalFuture<?> reduceValidatePartsAndFinish(
+        UUID procId,
+        Map<UUID, CheckResultDTO> results,
+        Map<UUID, Throwable> errors
+    ) {
+        clean(procId, null, results, errors);
+
+        return new GridFinishedFuture<>();
+    }
+
+    /** Phase 2 beginning.  */
+    private IgniteInternalFuture<CheckResultDTO> validateParts(SnapshotCheckProcessRequest req) {
+        if (req.error() != null)
+            return new GridFinishedFuture<>(req.error());
+
+        SnapshotCheckProcessRequest locReq = currentRequest(null, req.requestId());
+
+        if (locReq == null)
+            return new GridFinishedFuture<>();
+
+        assert locReq.equals(req);
+
+        // Store metas on the originator to redulte relusts later.
+        if (req.operationalNodeId().equals(kctx.localNodeId()))
+            locReq.metas = req.metas;
+
+        // Local meta might be null if current node started after the snapshot creation or placement.
+        if (!req.nodes.contains(kctx.localNodeId()) || locReq.meta() == null)
+            return new GridFinishedFuture<>();
+
+        GridFutureAdapter<CheckResultDTO> locPartsChkFut = new GridFutureAdapter<>();
+
+        stopFutureOnAnyFailure(locPartsChkFut, () -> {
+            locReq.fut(locPartsChkFut);
+
+            if (!locPartsChkFut.isDone()) {
```

Comment:
The same, but `checkPartitions` looks very heavy workload. How do you see if it possible to interrupt the processing in case of error?

Reply:
Yes, we should be able to interrupt here. But I see it as a new ticket like "Snapshot check cancelation." Currently, there is no normal snapshot check cancelation.